runtimes/runtimes/lsp/router/routerByServerName.ts

```diff
@@ -0,0 +1,57 @@
+import { EventIdentifier, FollowupIdentifier, NotificationHandler } from '../../../protocol'
+import { Encoding } from '../../encoding'
+
+type NotificationId = {
+    serverName: string
+    id: string
+}
+
+export class RouterByServerName<P extends Partial<EventIdentifier>, F extends FollowupIdentifier> {
+    constructor(
+        private serverName: string,
+        private encoding: Encoding
+    ) {}
+
+    send(sendHandler: (params: P) => Promise<void>, params: P) {
+        const attachServerName = (): P => {
+            const idObject = {
+                serverName: this.serverName,
+                id: params.id!,
+            }
+            const id = this.encoding.encode(JSON.stringify(idObject))
+            return {
+                ...params,
+                id,
+            }
+        }
+
+        const sendParams = params.id ? attachServerName() : params
+        sendHandler(sendParams)
+    }
+
+    processFollowup(followupHandler: NotificationHandler<F>, params: F) {
+        if (!params.source.id) {
+            return
+        }
+
+        const sourceId = this.encoding.decode(params.source.id)
+        const id = this.parseServerName(sourceId)
+        if (id?.serverName === this.serverName) {
```

Comment:
What if multiple servers have the same server name?

Reply:
There is check for this in `LspRouter`.